x-pack/platform/plugins/shared/onechat/server/routes/mcp.ts

```diff
@@ -0,0 +1,201 @@
+          reason: 'todo',
+        },
+      },
+      access: 'public',
+      summary: 'MCP server',
+      description: TECHNICAL_PREVIEW_WARNING,
+      options: {
+        tags: ['mcp'],
+        xsrfRequired: false,
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: { body: schema.object({}, { unknowns: 'allow' }) },
+        },
+      },
+      wrapHandler(async (_, request, response) => {
+        let transport: KibanaMcpHttpTransport | undefined;
+        let server: McpServer | undefined;
+
+        try {
+          transport = new KibanaMcpHttpTransport({ sessionIdGenerator: undefined, logger });
+
+          // Instantiate new MCP server upon every request, no session persistence
+          server = new McpServer({
+            name: MCP_SERVER_NAME,
+            version: MCP_SERVER_VERSION,
+          });
+
+          const { tools: toolService } = getInternalServices();
+
+          const registry = toolService.registry.asScopedPublicRegistry({ request });
+          const tools = await registry.list({});
+
+          // Expose tools scoped to the request
+          for (const tool of tools) {
+            server.tool(
+              tool.id,
+              tool.description,
+              tool.schema.shape,
+              async (args: { [x: string]: any }) => {
+                const toolResult = await tool.execute({ toolParams: args });
+                return {
+                  content: [{ type: 'text' as const, text: JSON.stringify(toolResult) }],
+                };
+              }
+            );
+          }
+
+          request.events.aborted$.subscribe(() => {
+            transport?.close().catch((error) => {
+              logger.error('MCP Server: Error closing transport', { error });
+            });
+            server?.close();
+          });
+
+          await server.connect(transport);
+
+          return await transport.handleRequest(request, response);
+        } catch (error) {
+          logger.error('MCP Server: Error handling request', { error });
+          try {
+            await transport?.close();
+          } catch (closeError) {
+            logger.error('MCP Server: Error closing transport during error handling', {
+              error: closeError,
+            });
+          }
+          if (server) {
+            try {
+              server.close();
+            } catch (closeError) {
+              logger.error('MCP Server: Error closing server during error handling', {
+                error: closeError,
+              });
+            }
+          }
+
+          logger.error('MCP Server: Error handling request', { error });
+          return response.customError({
+            statusCode: 500,
+            body: {
+              message: `Internal server error: ${error}`,
+              attributes: {
+                code: -32603,
```

Comment:
Nit: leave an inline comment for what this error code represents.
EDIT: and/or constantize? I see these are being used also in the kibana_mcp_http_transport

Reply:
Hmm the codes are copy-pasted from the official implementation https://github.com/modelcontextprotocol/typescript-sdk/blob/main/src/server/streamableHttp.ts (they also have them defined directly) I can try tracking down what they mean ...

Reply:
fixed